qiita_pet/handlers/study_handlers/sample_template.py

```diff
@@ -57,3 +61,50 @@ def post(self):
         else:
             raise HTTPError(400, 'Unknown sample template action: %s' % action)
         self.write(result)
+
+
+def build_sample_summary(study_id, user_id, visible=None):
```

Comment:
I don't think memoization will buy us that much - and I don't think that precomputing is the way to go in here. The main concern that I have from here is that this is loading all the templates, but in the mockup it doesn't look like you need all that information. If I understand correctly, the only parts that you need is know in which prep template each sample is present and all the metadata categories that are available.

Based on that, you can load that information first and then just load the metadata columns that the user want to show once it adds them to the table.

It is also unclear from the mockup how columns that are present in one prep template but not in another prep template are handled.